### PR TITLE
Disable `lr_scheduler.step()` in manual optimization

### LINKED ISSUE #6379: Don't `lr_scheduler.step()` in manual optimization
## 🚀 Feature
Currently, `lr_scheduler.step()` is called in both manual and automatic optimization. We should let users call `lr_scheduler.step()` manually in manual optimization for ultimate flexibility. Requested by @carmocca.

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #6379. Once #6567 and this PR are merged, users can call `lr_scheduler.step()` at arbitrary intervals in manual optimization for ultimate flexibility.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/optimization/test_manual_optimization.py [hunk 1]
```diff
@@ -1147,3 +1147,33 @@ def dis_closure():
...
+        limit_train_batches=1,
+        limit_test_batches=1,
+        limit_val_batches=1,
+    )
+
+    with patch("torch.optim.lr_scheduler.StepLR.step") as lr_step:
+        trainer.fit(model)
+
+    # If a lr scheduler inherits `torch.optim.lr_scheduler._LRScheduler`,
+    # `.step()` is called once when its instantiation.
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:1177]:
```suggestion
    # `.step()` is called once during its instantiation.
```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 2]
```diff
@@ -1147,3 +1147,33 @@ def dis_closure():
...
+    model = TestModel()
+    model.training_step_end = None
+    model.training_epoch_end = None
+
+    trainer = Trainer(
+        max_epochs=1,
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_test_batches=1,
+        limit_val_batches=1,
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:1170]: should this be at least two?
fast_dev_run=2 maybe?
```suggestion
        fast_dev_run=2,
```

File: tests/trainer/optimization/test_manual_optimization.py [hunk 3]
```diff
@@ -1147,3 +1147,33 @@ def dis_closure():
...
+
+
+def test_lr_scheduler_step_not_called(tmpdir):
+    """
+    Test `lr_scheduler.step()` is not called in manual optimization.
+    """
+    class TestModel(BoringModel):
+        def __init__(self):
+            super().__init__()
+            self.automatic_optimization = False
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:1159]: let's implement training_step here with a call to manual backward and optimizer step?

Comment (akihironitta) [reply tests/trainer/optimization/test_manual_optimization.py:1159]: Addressed!

File: tests/trainer/optimization/test_manual_optimization.py (hunk 2 above)
Comment (akihironitta) [reply tests/trainer/optimization/test_manual_optimization.py]: Yes, it should be. Thanks.

File: CHANGELOG.md [hunk 4]
```diff
@@ -102,6 +102,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `PyTorchProfiler` to use `torch.autograd.profiler.record_function` to record functions ([#6349](https://github.com/PyTorchLightning/pytorch-lightning/pull/6349))
 
 
+- Disabled `lr_schdeulder.step()` in manual optimization  ([#6825](https://github.com/PyTorchLightning/pytorch-lightning/pull/6825))
```
Comment (tchaton) [inline CHANGELOG.md:105]:
```suggestion
- Disabled `lr_scheduler.step()` in manual optimization  ([#6825](https://github.com/PyTorchLightning/pytorch-lightning/pull/6825))
```

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 5]
```diff
@@ -32,7 +32,7 @@ def update_learning_rates(self, interval: str, monitor_metrics=None):
             interval: either 'epoch' or 'step'.
             monitor_metrics: dict of possible values to monitor
         """
-        if not self.trainer.lr_schedulers:
+        if not self.trainer.lr_schedulers or not self.trainer.train_loop.automatic_optimization:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:35]: Let's update the doc with this change + add a warning if we detect the scheduler was on step or epoch, but never triggered.

File: docs/source/common/optimizers.rst [hunk 6]
```diff
@@ -44,6 +44,8 @@ to manually manage the optimization process. To do so, do the following:
 
 .. warning:: Before 1.2, ``optimzer.step`` was calling ``optimizer.zero_grad()`` internally. From 1.2, it is left to the users expertise.
 
+.. warning:: Before 1.3, ``lr_scheduler.step`` was called automatically in both manual and automatic optimization. From 1.3, ``lr_scheduler.step`` is disabled in manual optimization so that you can call it at arbitrary intervals. Use ``self.lr_schedulers()`` in LightningModule to access your learning rate schedulers defined in ``LightningModule.configure_optimizers()``.
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:47]: `self.lr_schedulers()` is being added in #6567.

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 5 above)
Comment (akihironitta) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:35]: I updated the docs and added a warning when scheduler keys (e.g. "interval", "frequency", "monitor"), which are invalid in manual optimization, are provided. Would you mind having a look at the changes?

**Review by awaelchli — APPROVED** [review]
If this breaking change is considered major, one could alternatively warn today that from version 1.x the behaviour will change, and until then keep the current implementation. Just a consideration.

File: pytorch_lightning/trainer/optimizers.py [hunk 7]
```diff
@@ -117,6 +123,16 @@ def configure_schedulers(self, schedulers: list, monitor: Optional[str] = None):
                         f'The "interval" key in lr scheduler dict must be "step" or "epoch"'
                         f' but is "{scheduler["interval"]}"'
                     )
+                if is_manual_optimization:
+                    invalid_keys = {'interval', 'frequency', 'reduce_on_plateau', 'monitor', 'strict'}
```
Comment (carmocca) [inline pytorch_lightning/trainer/optimizers.py:127]: why are reduce_on_plateau, monitor, and strict invalid?

File: pytorch_lightning/trainer/optimizers.py [hunk 8]
```diff
@@ -98,8 +99,13 @@ def _convert_to_lightning_optimizer(trainer, optimizer):
...
         }
 
-    def configure_schedulers(self, schedulers: list, monitor: Optional[str] = None):
-        # Convert each scheduler into dict structure with relevant information
+    def configure_schedulers(
+        self,
+        schedulers: list,
+        monitor: Optional[str],
+        is_manual_optimization: bool,
+    ):
```
Comment (carmocca) [inline pytorch_lightning/trainer/optimizers.py:107]: could take this chance to add a return type :)

File: pytorch_lightning/trainer/optimizers.py (hunk 7 above)
Comment (akihironitta) [reply pytorch_lightning/trainer/optimizers.py:127]: Because even if those keys are provided, users need to call `lr_scheduler.step(metric_to_monitor)` manually anyway since `lr_scheduler.step()` will be disabled for all learning rate schedulers including [`torch.optim.lr_scheduler.ReduceLROnPlateau`](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau) in manual optimization.
```python
scheduler = self.lr_schedulers()
scheduler.step(val_loss)
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:35]: Looks good. Could we add in the doc how to use the scheduler for interval=`epoch`.

```
 def training_step(self, batch, batch_idx)

     # do optimization

      if self.trainer.is_last_batch: # is_last_batch doesn't exist for the trainer, could be added.
             self.lr_schedulers[0].step()
```

**Review by tchaton — COMMENTED** [review]
LGTM ! Let's add example in the documentation how to call the schedulers on different internal

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 5 above)
Comment (akihironitta) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:35]: @tchaton Addressed! Could you have a look again?

**Comment by akihironitta** [comment]
In my opinion, [the current docs about optimization](https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html) (especially about manual optimization) are somewhat messy. There are quite a number of tip/note/warning which are randomly distributed. I'll submit a follow-up PR to clean up and restructure the whole optimization docs.

File: CHANGELOG.md [hunk 9]
```diff
@@ -102,6 +102,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `PyTorchProfiler` to use `torch.autograd.profiler.record_function` to record functions ([#6349](https://github.com/PyTorchLightning/pytorch-lightning/pull/6349))
 
 
+- Disabled `lr_scheduler.step()` in manual optimization  ([#6825](https://github.com/PyTorchLightning/pytorch-lightning/pull/6825))
```
Comment (tchaton) [inline CHANGELOG.md:105]: Can you add `is_last_batch` to the changelog.

File: docs/source/common/optimizers.rst [hunk 10]
```diff
@@ -44,6 +44,27 @@ to manually manage the optimization process. To do so, do the following:
...
+.. testcode:: python
+
+   def __init__(self):
+       self.automatic_optimization = False
+
+   def training_step(self, batch, batch_idx):
+       # do foward, backward, and optimization
+       ...
+
+       # step every `n` batches
```
Comment (tchaton) [inline docs/source/common/optimizers.rst:58]: Really neat example !

File: pytorch_lightning/trainer/training_loop.py [hunk 11]
```diff
@@ -413,6 +413,7 @@ def run_training_epoch(self):
         for batch_idx, (batch, is_last_batch) in train_dataloader:
 
             self.trainer.batch_idx = batch_idx
+            self.trainer.is_last_batch = is_last_batch
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:416]: Nice !

**Review by tchaton — APPROVED** [review]
Great work there !

**Comment by tchaton** [comment]
@akihironitta. There is a bug with Pipe.

Screenshot 2021-04-09 at 10 28 58

**Comment by akihironitta** [comment]
I think the failing tests on Windows are not related to the changes here as they're also failing on master.
```
tests/trainer/test_data_loading.py::test_dataloader_warnings[0] FAILED   [ 63%]
tests/trainer/test_data_loading.py::test_dataloader_warnings[1] FAILED   [ 63%]
```

**Review by Borda — CHANGES_REQUESTED** [review]
seems like all Windows tests are failing

**Comment by carmocca** [comment]
Windows errors are caused by https://github.com/PyTorchLightning/pytorch-lightning/pull/6762

**Comment by akihironitta** [comment]
@Borda The failing windows tests should be fixed now by @carmocca in #6930, so would you mind having a look again?

**Comment by akihironitta** [comment]
Just for the note, as #6907 has been merged already, there's no need to change any docs here, so I reverted the docs change.
